sdk/storage/azblob/lease/blob_client.go

```diff
@@ -0,0 +1,141 @@
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package lease
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/appendblob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blockblob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/internal/base"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/internal/generated"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/internal/shared"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/pageblob"
+)
+
+// BlobClient provides lease functionality for the underlying blob client.
+type BlobClient struct {
+	blobClient *blob.Client
+	leaseID    *string
+}
+
+// BlobClientOptions contains the optional values when creating a BlobClient.
+type BlobClientOptions struct {
+	// LeaseID contains a caller-provided lease ID.
+	LeaseID *string
+}
+
+// NewBlobClient creates a blob lease client for the provided blob client.
+func NewBlobClient[T appendblob.Client | blob.Client | blockblob.Client | pageblob.Client](client *T, options *BlobClientOptions) (*BlobClient, error) {
+	var leaseID *string
+	if options != nil {
+		leaseID = options.LeaseID
+	}
+
+	leaseID, err := shared.GenerateLeaseID(leaseID)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: improve once generics supports this scenario
+	var blobClient *blob.Client
+	switch t := any(client).(type) {
+	case *appendblob.Client:
+		rawClient, _ := base.InnerClients((*base.CompositeClient[generated.BlobClient, generated.AppendBlobClient])(t))
+		blobClient = (*blob.Client)(rawClient)
+	case *blockblob.Client:
+		rawClient, _ := base.InnerClients((*base.CompositeClient[generated.BlobClient, generated.BlockBlobClient])(t))
+		blobClient = (*blob.Client)(rawClient)
+	case *pageblob.Client:
+		rawClient, _ := base.InnerClients((*base.CompositeClient[generated.BlobClient, generated.PageBlobClient])(t))
+		blobClient = (*blob.Client)(rawClient)
+	case *blob.Client:
+		blobClient = t
+	default:
+		// this shouldn't happen due to the generic type constraint
+		return nil, fmt.Errorf("unhandled client type %T", client)
+	}
+
+	return &BlobClient{
+		blobClient: blobClient,
+		leaseID:    leaseID,
+	}, nil
+}
+
+func (c *BlobClient) generated() *generated.BlobClient {
+	return base.InnerClient((*base.Client[generated.BlobClient])(c.blobClient))
+}
+
+// LeaseID returns leaseID of the client.
+func (c *BlobClient) LeaseID() *string {
+	return c.leaseID
+}
+
+// AcquireLease acquires a lease on the blob for write and delete operations.
+// The lease Duration must be between 15 and 60 seconds, or infinite (-1).
+// For more information, see https://docs.microsoft.com/rest/api/storageservices/lease-blob.
+func (c *BlobClient) AcquireLease(ctx context.Context, o *BlobAcquireOptions) (BlobAcquireResponse, error) {
```

Comment:
I'm thinking we should remove the `Lease` suffix from these APIs as they're already on a type within the `lease` package.  Thoughts?

Reply:
I think we should keep it as it makes API names more readable. The function is built on client that may not have lease in the name. `client.Acquire()` is less readable than `client.AcquireLease()`.